server/src/main/java/org/apache/lucene/index/BaseSingleStarTreeBuilder.java

```diff
@@ -0,0 +1,660 @@
+        // CodecUtil.writeIndexHeader(indexOutput, STAR_TREE_CODEC, 0, state.segmentInfo.getId(), state.segmentSuffix);
+        this.mapperService = mapperService;
+        this.starTreeField = starTreeField;
+        this.starTreeFieldSpec = starTreeField.getStarTreeConfig();
+        this.docValuesConsumer = docValuesConsumer;
+        this.docValuesProducer = docValuesProducer;
+        this.starTreeDocValuesIteratorFactory = new StarTreeDocValuesIteratorFactory();
+
+        List<Dimension> dimensionsSplitOrder = starTreeField.getDimensionsOrder();
+        this.numDimensions = dimensionsSplitOrder.size();
+        this.dimensionsSplitOrder = new String[numDimensions];
+
+        this.skipStarNodeCreationForDimensions = new HashSet<>();
+        this.totalDocs = state.segmentInfo.maxDoc();
+        this.dimensionReaders = new DocIdSetIterator[numDimensions];
+        Set<String> skipStarNodeCreationForDimensions = this.starTreeFieldSpec.getSkipStarNodeCreationInDims();
+
+        for (int i = 0; i < numDimensions; i++) {
+            String dimension = dimensionsSplitOrder.get(i).getField();
+            this.dimensionsSplitOrder[i] = dimension;
+            if (skipStarNodeCreationForDimensions.contains(dimensionsSplitOrder.get(i).getField())) {
+                this.skipStarNodeCreationForDimensions.add(i);
+            }
+            FieldInfo dimensionFieldInfos = state.fieldInfos.fieldInfo(dimension);
+            DocValuesType dimensionDocValuesType = state.fieldInfos.fieldInfo(dimension).getDocValuesType();
+            dimensionReaders[i] = starTreeDocValuesIteratorFactory.createIterator(
+                dimensionDocValuesType,
+                dimensionFieldInfos,
+                docValuesProducer
+            );
+        }
+
+        this.metricStatFieldPairs = generateMetricStatFieldPairs();
+        this.numMetrics = metricStatFieldPairs.size();
+        this.metrics = new String[numMetrics];
+        this.valueAggregators = new ValueAggregator[numMetrics];
+        this.numericTypes = new IndexNumericFieldData.NumericType[numMetrics];
+        this.metricReaders = new DocIdSetIterator[numMetrics];
+
+        int index = 0;
+        for (MetricStatFieldPair metricStatFieldPair : metricStatFieldPairs) {
+            metrics[index] = metricStatFieldPair.toFieldName();
+            valueAggregators[index] = ValueAggregatorFactory.getValueAggregator(metricStatFieldPair.getMetricStat());
+
+            Mapper fieldMapper = mapperService.documentMapper().mappers().getMapper(metrics[index]);
+            if (fieldMapper instanceof NumberFieldMapper) {
+                numericTypes[index] = ((NumberFieldMapper) fieldMapper).fieldType().numericType();
+            } else {
+                numericTypes[index] = IndexNumericFieldData.NumericType.DOUBLE;
+            }
+            // Ignore the column for COUNT aggregation function
+            if (valueAggregators[index].getAggregationType() != MetricStat.COUNT) {
```

Comment:
Lets have special handling for count across different fields to count per star tree ?